File: day4/group_anagram.cpp
#include<algorithm>
#include<string>
#include<vector>
#include<map>

using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<string> temp (strs);
        map<string, vector<string>> hashMap;
        vector<vector<string>> result;

        for(int i = 0; i < strs.size(); ++i){
            sort(strs[i].begin(), strs[i].end());
        }

        for(int i = 0; i < strs.size(); ++i){
            hashMap[strs[i]].push_back(temp[i]);
        }
        for(map<string, vector<string>>::iterator  it = hashMap.begin(); it != hashMap.end(); ++it){
            result.push_back(it->second);
        }

        return result;
                
    }
};

File: solutions/day_004/group_anagram.dart
import 'dart:collection';

class Solution {
  List<List<String>> groupAnagrams(List<String> strs) {
    // create a map to store the sorted string as the key and the list of strings as the value
    // HashMap is used because it is faster than the Map class
    HashMap<String, List<String>> map = HashMap();
    for (String str in strs) {
      // split the string into a list of characters
      List<String> chars = str.split("");
      // sort the list of characters
      chars.sort();

      String sortedStr = chars.toString();
      // if the map does not contain the sorted string, add it to the map with an empty list as the value
      map.putIfAbsent(sortedStr, () => []);
      // add the string to the list of strings
      map[sortedStr]?.add(str);
    }

    return map.values.toList();
  










  }
}